### ISSUE TITLE
URI_SAFE_ATTRIBUTES breaking change

### ISSUE THREAD

**Issue by charlag** [cure53/DOMPurify#366]
Hi,
thanks for your awesome work on DOMPurify.

### Background & Context

We're in the process of upgrading from 1.0.8 to 2.0.2 and we noticed that email bodies we parse do not look correct. Turns out some style tags were removed.

### Bug

`URI_SAFE_ATTRIBUTES` now completely overwrites default atributes and not extends the set.

#### Input

```js
DOMPurify.sanitize('<div style="color: red">Test</div>', {
  ADD_URI_SAFE_ATTR: ['poster'], // for video element
})
```

#### Given output

```html
<div>Test</div>
```

#### Expected output

```html
<div style="color: red">Test</div>
```

New behaviour is fine maybe but I did not find it mentioned anywhere.
Seems like it was introduced in fe3bcc6b2a56ebaff0b73cc9e1629e224d39aa91 (1.0.11) and it is used by `_isValidAttribute`.

Thanks.

**Comment by cure53** [comment]
Eek, we will have a look, thanks for the report!

**Comment by cure53** [comment]
The problem should be right here:

https://github.com/cure53/DOMPurify/blob/master/src/purify.js#L317

It appears that the additional attributes are added to an empty set and not the existing default attribute names. Building a fix now.

**Comment by charlag** [comment]
Yes, and the same is for `ALLOWED_ATTR`, runnin ginto it right now.

**Comment by charlag** [comment]
I meant `ADD_ATTR`, sorry

**Comment by cure53** [comment]
I have trouble to see the problem with `ADD_ATTR` - do you have a PoC?

**Comment by cure53** [comment]
@charlag I would propose the following to be a valid fix:

```javascript
URI_SAFE_ATTRIBUTES = 'ADD_URI_SAFE_ATTR' in cfg 
    ? addToSet(DEFAULT_URI_SAFE_ATTRIBUTES, cfg.ADD_URI_SAFE_ATTR) 
    : addToSet({}, DEFAULT_URI_SAFE_ATTRIBUTES);
```

This would address your issue without interfering with #327. Wdyt?

**Comment by cure53** [comment]
This should do the trick, can you confirm? @charlag 

```javascript
    URI_SAFE_ATTRIBUTES =
      'ADD_URI_SAFE_ATTR' in cfg
        ? addToSet(URI_SAFE_ATTRIBUTES, cfg.ADD_URI_SAFE_ATTR)
        : addToSet({}, DEFAULT_URI_SAFE_ATTRIBUTES);
```

**Comment by charlag** [comment]
I think it should be

```js
    URI_SAFE_ATTRIBUTES =
      'ADD_URI_SAFE_ATTR' in cfg
        ? addToSet(DEFAULT_URI_SAFE_ATTRIBUTES, cfg.ADD_URI_SAFE_ATTR)
        : addToSet({}, DEFAULT_URI_SAFE_ATTRIBUTES);
```

I'm trying to reproduce my problem with attributes right now

**Comment by cure53** [comment]
No, this is gonna blow everything up and will collide with #327.

**Comment by charlag** [comment]
Ah yes, seems like you're right. The thing is, by default `URI_SAFE_ATTRIBUTES` are `DEFAULT_URI_SAFE_ATTRIBUTES`

Currently in `_parseConfig()`

```js
URI_SAFE_ATTRIBUTES = 'ADD_URI_SAFE_ATTR' in cfg ? addToSet({}, cfg.ADD_URI_SAFE_ATTR) : DEFAULT_URI_SAFE_ATTRIBUTES;
```

default ones need to be cloned then? Otherwise we will add `ADD_URI_SAFE_ATTR` to `DEFAULT_URI_SAFE_ATTRIBUTES`

**Comment by charlag** [comment]
Okay so the other issue is related to the changes in the SVG, not in the attributes, sorry.

We replace image src with `data:image/svg+xml;utf8<svg ... />` in `afterSanitizeElements` and not it is removed because `removeSVGAttr === true`. We could change a hook to `afterSanitizeAttributes` I guess

**Comment by cure53** [comment]
I think I have a good fix for the original problem, please check latest commit.

**Comment by cure53** [comment]
> Okay so the other issue is related to the changes in the SVG, not in the attributes, sorry.
> 
> We replace image src with `data:image/svg+xml;utf8<svg ... />` in `afterSanitizeElements` and it is removed because `removeSVGAttr === true`. We could change a hook to `afterSanitizeAttributes` I guess

Ah yes, this indeed triggers the current mXSS check. This should better be fixed with a hook or simply by encoding it properly - then the problem should be gone.

**Comment by cure53** [comment]
That was quick :D Does it work as expected now?

**Comment by charlag** [comment]
@cure53 how can I test these changes easily? I think it should work

**Comment by cure53** [comment]
Check out this file for testing: https://github.com/cure53/DOMPurify/blob/master/dist/purify.js

Our tests are green and the local tests look good as well.

**Comment by cure53** [comment]
Top, thanks! A new release will come very soon as well.

**Comment by NicoleG25** [comment]
Hi @cure53  ! I'm assuming the full fix was released in 2.0.7 in https://github.com/cure53/DOMPurify/commit/4e8af7b2c4a159b683d317e02c5cbddb86dc4a0e but could you tell me what are the vulnerable versions?

Thanks in advance ! :)

**Comment by cure53** [comment]
Heya, all versions below 2.0.7 have bypasses. The problem was a new attack technique that we only learned to fully understand over the course of several days..
